### PR TITLE
Fix StorageMap.Put for UInt160 and UInt256

### LINKED ISSUE #624: Poor developer experience with StorageMap.Get<T> and Put<T>
It is very easy to mismatch `StorageMap` `Get` and `Put` methods. In the code below, the call to `owners.Put` actually calls `owners.Put<UInt160>`, which automatically calls `StdLib.Serialize` on the provided `UInt160 owner` parameter. However, since the type parameter for `StorageMap.Get<T>` method can't be determined implicitly, the call to `owners.Get` below actually calls the non-generic method which in turn does NOT call `StdLib.Deserialize`. This leads to a hard-to-diagnose runtime error when the serialized `UInt160` value is cast w/o calling Deserialize.

Example code:

``` csharp
readonly StorageMap owners = new StorageMap(Storage.CurrentContext, nameof(owners));

public UInt160 GetOwner(BigInteger id)
{
    ByteString owner = owners.Get(id); // calls non-generic Get method, which doesn't call StdLib.Deserialize
    return (UInt160)owner ?? UInt160.Zero; // this cast fails at runtime if the owner is not null
}

public void PutOwner(BigInteger id, UInt160 owner)
{
    owners.Put(id, owner); // implicitly calls StorageMap.Put<UInt160> which calls StdLib.Serialize on the provided parameter
}
```

### PR DESCRIPTION
**Pull request by erikzhang**
Fix #624

### CODE REVIEW

**Review by shargon — APPROVED** [review]
Code looks good, please @devhawk confirm that the issue was solved

**Comment by devhawk** [comment]
This solves #624 for UInt160/256, but it doesn't solve the general problem. The core issue is that Put<T> is called *implicitly* while Get<T> has to be called *explicitly*. Imagine the same scenario with a user defined struct:

``` csharp

public class SaleInfo
{
    public ByteString Id;
    public UInt160 Seller;
    public UInt160 Buyer;
    public string Description;
    public BigInteger Price;
}

public class SafePurchaseContract : SmartContract
{
    readonly StorageMap salesMap = new StorageMap(Storage.CurrentContext, "saleInfo");

    public void CreateSale(ByteString saleId, BigInteger price, string description)
    {
        
        var tx = (Transaction)Runtime.ScriptContainer;
        var saleInfo = new SaleInfo()
        {
            Id = saleId,
            Seller = tx.Sender,
            Description = description,
            Price = price,
        };

        salesMap.Put(saleId, saleInfo); // Put<SaleInfo> implicitly called here
    }

    public string GetSale(ByteString saleId)
    {
        var sale = salesMap.Get<SaleInfo>(saleId); // Get<SaleInfo> must be explicitly called here
        return StdLib.JsonSerialize(sale);
    }
}
```

**Comment by erikzhang** [comment]
@devhawk Do you have any idea? Rename the method to `PutSerialized`?

**Comment by devhawk** [comment]
> @devhawk Do you have any idea? Rename the method to `PutSerialized`?

Yes, I think rename `Put<T>` is the best solution. I like `Serialize` a little better than `PutSerialized` because it's only one word and it's shorter to type. But `PutSerialized` would show up directly next `Put` in the Intellisense list. Frankly, I don't have strong feelings about the new name as long as it's different.

We should also change the deserializing version of `Get<T>` to match - i.e. either `Deserialize` or `GetDeserialized`.

**Comment by erikzhang** [comment]
@devhawk Please review it again.

File: src/Neo.SmartContract.Framework/Services/StorageMap.cs [hunk 1]
```diff
@@ -105,14 +105,14 @@ public class StorageMap
         [Syscall("System.Storage.Get")]
         public extern ByteString Get(byte[] key);
 
-        public T Get<T>(ByteString key) where T : class, new()
+        public object GetObject(ByteString key)
```
Comment (devhawk) [inline src/Neo.SmartContract.Framework/Services/StorageMap.cs:108]: Do we need separate overloads for ByteString and byte[]?

File: src/Neo.SmartContract.Framework/UInt160.cs [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 namespace Neo
 {
-    public class UInt160 : ByteString
+    public abstract class UInt160 : ByteString
```
Comment (devhawk) [inline src/Neo.SmartContract.Framework/UInt160.cs:5]: Do we still need to mark these as abstract?

File: src/Neo.SmartContract.Framework/UInt256.cs [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 namespace Neo
 {
-    public class UInt256 : ByteString
+    public abstract class UInt256 : ByteString
```
Comment (devhawk) [inline src/Neo.SmartContract.Framework/UInt256.cs:5]: DO we need to mark this as abstract?

**Review by devhawk — APPROVED** [review]
Minor nitpicks, otherwise looks good

**Pull request merged**